File: src/main/java/org/kite9/diagram/visualization/compaction/slideable/MinimizeCompactionStep.java
package org.kite9.diagram.visualization.compaction.slideable;

import java.util.Set;

import org.kite9.diagram.common.algorithms.so.Slideable;
import org.kite9.diagram.common.objects.OPair;
import org.kite9.diagram.model.Connected;
import org.kite9.diagram.model.Connection;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.Rectangular;
import org.kite9.diagram.model.style.DiagramElementSizing;
import org.kite9.diagram.visualization.compaction.Compaction;
import org.kite9.diagram.visualization.compaction.segment.Segment;
import org.kite9.diagram.visualization.display.CompleteDisplayer;

/**
 * In some sorted order, minimizes the size of Rectangular elements in the
 * diagram.
 * 
 * @see ContainedElementSizeOptimisationStep
 * @see LinkLengthReductionOptimisationStep
 * 
 * @author robmoffat
 * 
 */
public class MinimizeCompactionStep extends AbstractSizingCompactionStep {

	public MinimizeCompactionStep(CompleteDisplayer cd) {
		super(cd);
	}

	@Override
	public boolean filter(Rectangular r) {
		return (r instanceof Container) && (((Container) r).getSizing() == DiagramElementSizing.MINIMIZE) ;
	}

	public void performSizing(Rectangular r, Compaction c, boolean horizontal) {
		SegmentSlackOptimisation hsso = c.getHorizontalSegmentSlackOptimisation();
		OPair<Slideable<Segment>> hs = hsso.getSlideablesFor(r);
		SegmentSlackOptimisation vsso = c.getVerticalSegmentSlackOptimisation();
		OPair<Slideable<Segment>> vs = vsso.getSlideablesFor(r);
		if ((hs != null) && (vs != null)) {
			log.send("Minimizing Distance " + r);
			if (horizontal) {
				minimizeDistance(vsso, vs.getA(), vs.getB());
			} else {
				minimizeDistance(hsso, hs.getA(), hs.getB());
			}
		}
	}

	private int minimizeDistance(SegmentSlackOptimisation opt, Slideable<Segment> from, Slideable<Segment> to) {
		Integer minDist = from.minimumDistanceTo(to);
		opt.ensureMaximumDistance(from, to, minDist);
		
		return minDist;
	}

	public String getPrefix() {
		return "MINC";
	}

	public boolean isLoggingEnabled() {
		return true;
	}

	/**
	 * Returns in an order to maximize number of centerings.
	 */
	public int compare(Rectangular a, Rectangular b, Compaction c, boolean horizontal) {
		if (a.getDepth() != b.getDepth()) {
			return -((Integer) a.getDepth()).compareTo(b.getDepth());
		}

		if ((!(a instanceof Connected)) && (!(b instanceof Connected))) {
			return 0;
		} else if (!(a instanceof Connected)) {
			return -1;
		} else if (!(b instanceof Connected)) {
			return 1;
		} else {
			// return elements with least number of connections on a side
			int ac = maxLeavings(a, c, horizontal);
			int bc = maxLeavings(b, c, horizontal);
			if (bc != ac) {
				return ((Integer) ac).compareTo(bc);
			}
		}

		return b.getID().compareTo(a.getID());
	}

	private int maxLeavings(Rectangular a, Compaction c, boolean horizontal) {
		if (horizontal) {
			return maxLeavingsInAxis(a, c.getHorizontalSegmentSlackOptimisation(), c);
		} else {
			return maxLeavingsInAxis(a, c.getVerticalSegmentSlackOptimisation(), c);
		}
	}

	private int maxLeavingsInAxis(Rectangular a, SegmentSlackOptimisation sso, Compaction c) {
		OPair<Slideable<Segment>> along = sso.getSlideablesFor(a);
		return Math.max(leavingsOnSide(along.getA(), c), leavingsOnSide(along.getB(), c));
	}

	private int leavingsOnSide(Slideable<Segment> a2, Compaction c) {
		Set<Connection> connections = getLeavingConnections(a2.getUnderlying(), c);
		return connections.size();
	}

}

File: k9layout/src/main/kotlin/org/kite9/diagram/visualization/compaction/slideable/MinimizeCompactionStep.kt
package org.kite9.diagram.visualization.compaction.slideable

import org.kite9.diagram.common.algorithms.so.Slideable
import org.kite9.diagram.model.Connected
import org.kite9.diagram.model.Rectangular
import org.kite9.diagram.model.SizedRectangular
import org.kite9.diagram.model.style.DiagramElementSizing
import org.kite9.diagram.visualization.compaction.Compaction
import org.kite9.diagram.visualization.compaction.segment.Segment
import org.kite9.diagram.visualization.display.CompleteDisplayer
import kotlin.math.max

/**
 * In some sorted order, minimizes the size of Rectangular elements in the
 * diagram.
 *
 * @author robmoffat
 */
class MinimizeCompactionStep(cd: CompleteDisplayer?) : AbstractSizingCompactionStep(cd!!) {
    override fun filter(r: Rectangular, horizontal: Boolean): Boolean {
        return r is SizedRectangular && r.getSizing(horizontal) === DiagramElementSizing.MINIMIZE
    }

    override fun performSizing(r: Rectangular, c: Compaction, horizontal: Boolean) {
        val hsso = c.getHorizontalSegmentSlackOptimisation()
        val hs = hsso.getSlideablesFor(r)
        val vsso = c.getVerticalSegmentSlackOptimisation()
        val vs = vsso.getSlideablesFor(r)
        log.send("Minimizing Distance $r")
        if (horizontal) {
            minimizeDistance(vsso, vs.a!!, vs.b!!)
        } else {
            minimizeDistance(hsso, hs.a!!, hs.b!!)
        }

    }

    private fun minimizeDistance(
        opt: SegmentSlackOptimisation,
        from: Slideable,
        to: Slideable
    ): Int {
        val minDist = from.minimumDistanceTo(to)
        opt.ensureMaximumDistance(from, to, minDist)
        return minDist
    }

    override val prefix: String
        get() = "MINC"

    override val isLoggingEnabled: Boolean
        get() = true

    /**
     * Returns in an order to maximize number of centerings.
     */
    override fun compare(a: Rectangular, b: Rectangular, c: Compaction, horizontal: Boolean): Int {
        if (a.getDepth() != b.getDepth()) {
            return -a.getDepth().compareTo(b.getDepth())
        }
        if (a !is Connected && b !is Connected) {
            return 0
        } else if (a !is Connected) {
            return -1
        } else if (b !is Connected) {
            return 1
        } else {
            // return elements with least number of connections on a side
            val ac = maxLeavings(a, c, horizontal)
            val bc = maxLeavings(b, c, horizontal)
            if (bc != ac) {
                return ac.compareTo(bc)
            }
        }
        return b.getID().compareTo(a.getID())
    }

    private fun maxLeavings(a: Rectangular, c: Compaction, horizontal: Boolean): Int {
        return if (horizontal) {
            maxLeavingsInAxis(a, c.getHorizontalSegmentSlackOptimisation(), c)
        } else {
            maxLeavingsInAxis(a, c.getVerticalSegmentSlackOptimisation(), c)
        }
    }

    private fun maxLeavingsInAxis(a: Rectangular, sso: SegmentSlackOptimisation, c: Compaction): Int {
        val (a1, b) = sso.getSlideablesFor(a)
        return max(leavingsOnSide(a1!!, c), leavingsOnSide(b!!, c))
    }

    private fun leavingsOnSide(a2: Slideable, c: Compaction): Int {
        val connections = getLeavingConnections(a2.underlying, c)
        return connections.size
    }
}